Spigot-Server-Patches/0616-MC-4-Fix-item-position-desync.patch

```diff
@@ -0,0 +1,84 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: BillyGalbreath <blake.galbreath@gmail.com>
+Date: Tue, 8 Dec 2020 20:24:52 -0600
+Subject: [PATCH] MC-4: Fix item position desync
+
+This fixes item position desync (MC-4) by running the item coordinates
+through the encode/decode methods of the packet that causes the precision
+loss, which forces the server to lose the same precision as the client
+keeping them in sync.
+
+diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+index cdf1177ca7b4b1b0808176c455f62b395146da50..26db8a5d92d494e5f6ba9498d5a1aa555642a92b 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+@@ -469,4 +469,9 @@ public class PaperConfig {
+     private static void trackPluginScoreboards() {
+         trackPluginScoreboards = getBoolean("settings.track-plugin-scoreboards", false);
+     }
++
++    public static boolean fixEntityPositionDesync = true;
++    private static void fixEntityPositionDesync() {
++        fixEntityPositionDesync = getBoolean("settings.fix-entity-position-desync", fixEntityPositionDesync);
++    }
+ }
+diff --git a/src/main/java/net/minecraft/server/EntityItem.java b/src/main/java/net/minecraft/server/EntityItem.java
+index ba73d14437cfdf07ef0f1f6266131c113c2741fd..f41aaa7623c052b9f4044898d1bdee898c03057a 100644
+--- a/src/main/java/net/minecraft/server/EntityItem.java
++++ b/src/main/java/net/minecraft/server/EntityItem.java
+@@ -523,4 +523,16 @@ public class EntityItem extends Entity {
+     public Packet<?> P() {
+         return new PacketPlayOutSpawnEntity(this);
+     }
++
++    // Paper start - fix MC-4
++    public void setPositionRaw(double x, double y, double z) {
++        if (com.destroystokyo.paper.PaperConfig.fixEntityPositionDesync) {
++            // encode/decode from PacketPlayOutEntity
++            x = MathHelper.floorLong(x * 4096.0D) * (1 / 4096.0D);
```

Comment:
hmm just noticed this is flooring x/y/z, this would lock you at flat x/y/z so cant be mid block etc. that def was source of the issue for players.

But I question that overall for items even... can figure out how to reduce precision without throwing every decimal place away?

Reply:
Its not throwing them all away. I forget how many it leaves behind, but it is definitely smaller than the eye can see. You end up with like 4 or 5 decimals instead of 20 (or whatever). Thats what I meant about the precision loss

Reply:
doh my brain just took x + floor together and forgot about the 4096 modifier. ok so ignore that comment. 

I think this should be fine then. I question if we maybe should of just excluded players but I suspect it could have also had issues with entity pathing still.